File: skellig-connection-cassandra/src/main/java/org/skellig/connection/cassandra/CassandraRequestExecutor.java
package org.skellig.connection.cassandra;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.PlainTextAuthProvider;
import com.datastax.driver.core.Session;
import org.skellig.connection.cassandra.model.CassandraDetails;
import org.skellig.connection.database.DatabaseRequestExecutor;
import org.skellig.connection.database.model.DatabaseRequest;

public class CassandraRequestExecutor implements DatabaseRequestExecutor {

    private CassandraRequestExecutorFactory factory;
    private Cluster cluster;
    private Session session;

    public CassandraRequestExecutor(CassandraDetails cassandraDetails) {
        Cluster.Builder clusterBuilder = Cluster.builder()
                .addContactPointsWithPorts(cassandraDetails.getNodes());
        if (cassandraDetails.getUsername().isPresent()) {
            clusterBuilder.withAuthProvider(
                    new PlainTextAuthProvider(cassandraDetails.getUsername().get(), cassandraDetails.getPassword().orElse(null)));
        }
        cluster = clusterBuilder.build();
        session = cluster.connect();

        factory = new CassandraRequestExecutorFactory(session);
    }

    @Override
    public Object execute(DatabaseRequest request) {
        return factory.get(request).execute(request);
    }

    @Override
    public void close() {
        session.close();
        cluster.close();
    }
}

File: skellig-test-step-processing-cassandra/src/main/kotlin/org/skellig/teststep/processor/cassandra/CassandraRequestExecutor.kt
package org.skellig.teststep.processor.cassandra

import com.datastax.oss.driver.api.core.CqlSession
import org.skellig.teststep.processing.util.debug
import org.skellig.teststep.processing.util.logger
import org.skellig.teststep.processor.cassandra.model.CassandraDetails
import org.skellig.teststep.processor.db.DatabaseRequestExecutor
import org.skellig.teststep.processor.db.model.DatabaseRequest

/**
 * Represents a Cassandra request executor for executing database requests.
 * It connects to Cassandra DB when object is created.
 *
 * @property cassandraDetails The Cassandra details for connecting to the database server.
 */
class CassandraRequestExecutor(private val cassandraDetails: CassandraDetails) : DatabaseRequestExecutor {

    private val log = logger<CassandraRequestExecutor>()
    private val sessionInit = lazy {
        log.debug { "Initialize Cassandra Request Executor and Session for server '${cassandraDetails.serverName}'" }
        val sessionBuilder = CqlSession.builder()
            .withLocalDatacenter(cassandraDetails.datacenter ?: "datacenter1")
            .addContactPoints(cassandraDetails.nodes)
        cassandraDetails.userName?.let {
            sessionBuilder.withAuthCredentials(it, cassandraDetails.password ?: "")
        }
        sessionBuilder.build()
    }
    private val session: CqlSession by sessionInit
    private val factory: CassandraRequestExecutorFactory by lazy {
        CassandraRequestExecutorFactory(session)
    }

    override fun execute(databaseRequest: DatabaseRequest): Any? {
        return factory[databaseRequest]?.execute(databaseRequest)
    }

    /**
     * Closes the session associated with the CassandraRequestExecutor.
     * It should be called after executing a database request and when the CassandraRequestExecutor is no longer needed.
     */
    override fun close() {
        if (sessionInit.isInitialized()) {
            log.debug("Close Database connection to $cassandraDetails")
            session.close()
        }
    }
}